http-client/src/main/java/io/micronaut/http/client/netty/DefaultHttpClient.java

```diff
@@ -1491,181 +1454,360 @@ private <I> Publisher<URI> resolveURI(io.micronaut.http.HttpRequest<I> request,
+    /**
+     * This is the low-level request method, without redirect handling and with raw body bytes.
+     *
+     * @param poolHandle         The pool handle to send the request on
+     * @param request            The request to send
+     * @param byteBody           The request body
+     * @param mapConnectionError Operator used to map connection errors (e.g. when the connection
+     *                           is closed by the remote). Other errors are not mapped
+     * @return A mono containing the response
+     */
+    private Mono<NettyClientByteBodyResponse> sendRawRequest(
+        ConnectionManager.PoolHandle poolHandle,
+        io.micronaut.http.HttpRequest<?> request,
+        NettyByteBody byteBody,
+        UnaryOperator<Throwable> mapConnectionError
+    ) {
+        URI uri = request.getUri();
+        String uriWithoutHost = uri.getRawPath();
+        if (uri.getRawQuery() != null) {
+            uriWithoutHost += "?" + uri.getRawQuery();
+        }
+        HttpRequest nettyRequest = NettyHttpRequestBuilder.asBuilder(request)
+            .toHttpRequestWithoutBody()
+            .setUri(uriWithoutHost);
+
+        return Mono.<NettyClientByteBodyResponse>create(sink -> {
+            if (log.isDebugEnabled()) {
+                log.debug("Sending HTTP {} to {}", request.getMethodName(), request.getUri());
+            }
 
-        // build the mock netty request (only the content-type matters)
-        HttpRequest nettyRequest = new DefaultHttpRequest(HttpVersion.HTTP_1_1, HttpMethod.GET, "/");
-        List<AsciiString> relevantHeaders = List.of(HttpHeaderNames.CONTENT_TYPE);
-        for (AsciiString header : relevantHeaders) {
-            nettyRequest.headers().add(header, request.getHeaders().getAll(header));
-        }
+            boolean expectContinue = HttpUtil.is100ContinueExpected(nettyRequest);
+            ChannelPipeline pipeline = poolHandle.channel.pipeline();
 
-        HttpPostRequestEncoder encoder = buildMethod.apply(nettyRequest);
-        HttpRequest finalized = encoder.finalizeRequest();
-        // copy back the content-type
-        for (AsciiString header : relevantHeaders) {
-            request.getHeaders().remove(header);
-            for (String value : finalized.headers().getAll(header)) {
-                request.getHeaders().add(header, value);
+            // if the body is streamed, we have a StreamWriter, otherwise we have a ByteBuf.
+            StreamWriter streamWriter;
+            ByteBuf byteBuf;
+            if (byteBody instanceof AvailableNettyByteBody available) {
+                byteBuf = AvailableNettyByteBody.toByteBuf(available);
+                streamWriter = null;
+            } else {
+                streamWriter = new StreamWriter((StreamingNettyByteBody) byteBody, e -> {
+                    poolHandle.taint();
+                    sink.error(e);
+                });
+                pipeline.addLast(streamWriter);
+                byteBuf = null;
```

Comment:
Intellij noticed that byteBuf can be null but the Netty API doesn't allow it to be null

Reply:
all uses of byteBuf are guarded with checks of either byteBuf or streamWriter